matlab/FeatherReader.cpp

```diff
@@ -0,0 +1,301 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <arrow/io/file.h>
+#include <arrow/ipc/feather.h>
+#include <arrow/status.h>
+#include <arrow/table.h>
+#include <arrow/type.h>
+
+#include <mex.h>
+
+#include "FeatherReader.h"
+#include "FeatherType.h"
+
+namespace matlab {
+namespace arrow {
+namespace ipc {
+namespace feather {
+
+// MATLAB Arrays cannot be larger than 2^48
+static const uint64_t MAX_MATLAB_SIZE = static_cast<uint64_t>(0x01) << 48;
+
+// FeatherReader Constructor.
+FeatherReader::FeatherReader(const std::string& filename) : filename_{filename} {
+  // Open the given Feather file for reading.
+  Open();
```

Comment:
As a rule, elsewhere in the codebase we don't do any work in constructors (https://google.github.io/styleguide/cppguide.html#Doing_Work_in_Constructors). Since you're signaling errors that can fail to the mex runtime I would say this is an exception, but if there is a way to avoid a non-trivial constructor that would make things more consistent across the codebase